src/window-provider.ts

```diff
@@ -7,42 +9,110 @@ const WINDOW_PROVIDER_TARGET = "tally-window-provider"
 const unsafePostMessage = window.postMessage
 const unsafeAddEventListener = window.addEventListener
+const unsafeRemoveEventListener = window.removeEventListener
 const unsafeOrigin = window.location.origin
 
-// to demonstrate how it works it was necessary. Will remove later
-// eslint-disable-next-line no-console
-console.log("inpage.js in da house", Date.now())
-
 const WINDOW_PROVIDER_FLAG = "isTallyWindowProviderEnabled"
 
 const enabled = window.localStorage.getItem(WINDOW_PROVIDER_FLAG)
 
-if (enabled === "true") {
-  setInterval(() => {
-    // ‼️ Always include target origin to avoid unwanted attention
-    unsafePostMessage(
-      {
-        target: PROVIDER_BRIDGE_TARGET,
-        message: "SYN",
+class TallyWindowProvider extends EventEmitter {
+  // TODO: This should come from the background with onConnect when any interaction is initiated by the dApp.
+  // onboard.js relies on this, or uses a deprecated api. It seemed to be a reasonable workaround for now.
+  chainId: number | undefined = 1
+
+  selectedAddress: string | undefined
+
+  isConnected = false
+
+  isTally = true
+
+  bridgeListeners = new Map()
+
+  async request(request: {
+    method: string
+    params?: Array<unknown>
+  }): Promise<unknown> {
+    return this.send(request.method, request.params || [])
+  }
+
+  async send(method: string, params?: Array<unknown>): Promise<unknown> {
+    const sendData = {
+      id: Date.now().toString(),
+      target: PROVIDER_BRIDGE_TARGET,
+      payload: {
+        method,
+        params,
       },
-      unsafeOrigin
-    )
-
-    // to demonstrate how it works it was necessary. Will remove later
-    // eslint-disable-next-line no-console
-    console.log("-------------")
-  }, 1000)
-
-  unsafeAddEventListener("message", (event) => {
-    if (
-      event.origin !== unsafeOrigin || // we want to recieve msgs only from the provider bridge
-      event.source !== window || // we want to recieve msgs only from the provider bridge
-      event.data.target !== WINDOW_PROVIDER_TARGET
-    )
-      return
-
-    // to demonstrate how it works it was necessary. Will remove later
-    // eslint-disable-next-line no-console
-    console.log("inpage: ", event.data)
-  })
+    }
+
+    // ‼️ Always include target origin to avoid unwanted attention
+    unsafePostMessage(sendData, unsafeOrigin)
+
+    return new Promise((resolve) => {
+      function listener(
+        this: TallyWindowProvider,
+        event: {
+          origin: string
+          source: unknown
+          data: { id: string; target: string; payload: { result: unknown } }
+        }
+      ) {
+        if (
+          event.origin !== unsafeOrigin || // we want to recieve messages only from the provider-bridge script
+          event.source !== window || // we want to recieve messages only from the provider-bridge script
+          event.data.target !== WINDOW_PROVIDER_TARGET
+        ) {
+          return
+        }
```

Comment:
I like early returns in general, but I wonder if this would be easier to understand if we inverted the condition to a series of conjunctive affirmatives instead of a series of disjunctive negatives. I always find the latter harder to reason about. What that might look like:

```suggestion
        // Only handle messages (claiming to be) from the provider bridge.
        if (
          // only messages (claiming to be) from the same origin
          event.origin === unsafeOrigin &&
          // only messages (claiming to be) from the same window
          event.source === window &&
          // only messages (claiming to be) for the window provider
          event.data.target === WINDOW_PROVIDER_TARGET
        ) {
          // ... the rest of the code ...
        }
```